File: frontend/src/artifactNaming.ts
export function versionedBuildName(prefix: string, existingNames: string[] = [], date = new Date()): string {
  const pad = (value: number) => String(value).padStart(2, "0");
  const highestVersion = existingNames.reduce((highest, name) => {
    const match = name.match(/_v(\d+)$/i);
    return match ? Math.max(highest, Number(match[1])) : highest;
  }, 0);
  const version = Math.max(existingNames.length, highestVersion) + 1;
  return `${prefix}_${pad(date.getMonth() + 1)}${pad(date.getDate())}_v${String(version).padStart(3, "0")}`;
}

File: backend/app/artifact_naming.py
from __future__ import annotations

import re
from datetime import datetime
from typing import Iterable


def versioned_artifact_name(prefix: str, existing_names: Iterable[str], *, now: datetime | None = None) -> str:
    names = [str(name) for name in existing_names]
    highest_version = 0
    for name in names:
        match = re.search(r"_v(\d+)$", name, re.IGNORECASE)
        if match:
            highest_version = max(highest_version, int(match.group(1)))
    version = max(len(names), highest_version) + 1
    date = now or datetime.now()
    return f"{prefix}_{date.strftime('%m%d')}_v{version:03d}"
